flang/lib/Lower/OpenMP.cpp

```diff
@@ -714,6 +714,163 @@ static void genOMP(Fortran::lower::AbstractConverter &converter,
+  const auto &assignmentStmtExpr =
+      std::get<Fortran::parser::Expr>(std::get<3>(atomicWrite.t).statement.t);
+  const auto &assignmentStmtVariable = std::get<Fortran::parser::Variable>(
+      std::get<3>(atomicWrite.t).statement.t);
+  Fortran::lower::StatementContext stmtCtx;
+  auto value = fir::getBase(converter.genExprValue(
+      *Fortran::semantics::GetExpr(assignmentStmtExpr), stmtCtx));
+  if (auto varDesignator = std::get_if<
+          Fortran::common::Indirection<Fortran::parser::Designator>>(
+          &assignmentStmtVariable.u)) {
+    if (const auto *name = getDesignatorNameIfDataRef(varDesignator->value())) {
+      address = converter.getSymbolAddress(*name->symbol);
+    }
+  }
+  for (const auto &clause : leftHandClauseList.v) {
+    if (!hint)
+      hint = handleHintClause(clause);
+    if (memory_order == NULL)
+      memory_order = handleOmpAtomicMemoryOrderClause(converter, clause);
+  }
+
+  for (const auto &clause : rightHandClauseList.v) {
+    if (!hint)
+      hint = handleHintClause(clause);
+    if (memory_order == NULL) {
+      memory_order = handleOmpAtomicMemoryOrderClause(converter, clause);
+    }
+  }
+
+  firOpBuilder.create<mlir::omp::AtomicWriteOp>(currentLocation, address, value,
+                                                hint, memory_order);
+}
+
+static void genOmpAtomicRead(Fortran::lower::AbstractConverter &converter,
+                             Fortran::lower::pft::Evaluation &eval,
+                             const Fortran::parser::OmpAtomicRead &atomicRead) {
+  auto &firOpBuilder = converter.getFirOpBuilder();
+  auto currentLocation = converter.getCurrentLocation();
+  mlir::Type resultType;
+  mlir::Value address;
+  // If no hint clause is specified, the effect is as if
+  // hint(omp_sync_hint_none) had been specified.
+  uint64_t hint = 0;
+  mlir::StringAttr memory_order = NULL;
+  const auto &rightHandClauseList = std::get<2>(atomicRead.t);
+  const auto &leftHandClauseList = std::get<0>(atomicRead.t);
+  const auto &assignmentStmtExpr =
+      std::get<Fortran::parser::Expr>(std::get<3>(atomicRead.t).statement.t);
+  if (auto exprDesignator = std::get_if<
+          Fortran::common::Indirection<Fortran::parser::Designator>>(
+          &assignmentStmtExpr.u)) {
+    if (const auto *name =
+            getDesignatorNameIfDataRef(exprDesignator->value())) {
+      address = converter.getSymbolAddress(*name->symbol);
+      resultType = converter.genType(*name->symbol);
+    }
+  }
+  for (const auto &clause : rightHandClauseList.v) {
+    if (!hint)
+      hint = handleHintClause(clause);
+    if (memory_order == NULL) {
+      memory_order = handleOmpAtomicMemoryOrderClause(converter, clause);
+    }
+  }
+
+  for (const auto &clause : leftHandClauseList.v) {
+    if (!hint)
+      hint = handleHintClause(clause);
+    if (memory_order == NULL) {
+      memory_order = handleOmpAtomicMemoryOrderClause(converter, clause);
+    }
+  }
```

Comment:
The code in lines 835-849 are the same as above. Can we move it in one function?